lib/picos_sync/event.ml

```diff
@@ -0,0 +1,79 @@
+open Picos
+
+type 'a request = {
+  request : 'r. (unit -> 'r) Computation.t -> ('a -> 'r) -> unit;
+}
+[@@unboxed]
+
+type 'a event =
+  | Request : 'a request -> 'a event
+  | Choose : 'a event list -> 'a event
+  | Wrap : { event : 'b event; fn : 'b -> 'a } -> 'a event
+
+type 'a t = 'a event
+
+let[@inline] choose ts = Choose ts
+let[@inline] wrap event fn = Wrap { event; fn }
+let[@inline] map fn event = Wrap { event; fn }
+
+let rec request_1_as :
+      'a 'r. (unit -> 'r) Computation.t -> ('a -> 'r) -> 'a t -> unit =
+ fun computation to_result -> function
+  | Request { request } -> request computation to_result
+  | Choose ts -> request_n_as computation to_result ts
+  | Wrap { event; fn } ->
+      request_1_as computation (fun x -> to_result (fn x)) event
+
+and request_n_as computation to_result = function
+  | [] -> ()
+  | t :: ts ->
+      request_1_as computation to_result t;
+      request_n_as computation to_result ts
+
+let guard create_event =
+  let request computation to_result =
+    request_1_as computation to_result (create_event ())
+  in
+  Request { request }
+
+let await computation =
+  if Computation.is_running computation then begin
+    let t = Trigger.create () in
+    if Computation.try_attach computation t then
+      match Trigger.await t with
+      | None -> ()
+      | Some exn_bt ->
+          if Computation.try_cancel computation exn_bt then Exn_bt.raise exn_bt
+  end;
+  Computation.await computation ()
+
+let rec request_1 computation = function
+  | Request { request } -> request computation Fun.id
+  | Choose ts -> request_n computation ts
+  | Wrap { event; fn } -> request_1_as computation fn event
+
+and request_n computation = function
+  | [] -> ()
+  | t :: ts ->
+      request_1 computation t;
+      request_n computation ts
+
+let sync t =
+  let computation = Computation.create ~mode:`LIFO () in
+  match request_1 computation t with
+  | () -> await computation
+  | exception exn ->
+      let exn_bt = Exn_bt.get exn in
+      Computation.cancel computation exn_bt;
+      Exn_bt.raise exn_bt
+
+let select ts =
+  let computation = Computation.create ~mode:`LIFO () in
```

Comment:
Explicitly using `LIFO` ordering here, because the computation pretty much acts as a resource cleanup mechanism.